lib/Optimizer/GraphOptimizer/FunctionPass.cpp

```diff
@@ -0,0 +1,156 @@
+    llvm::cl::Optional, llvm::cl::cat(functionPassCat));
+
+llvm::cl::list<std::string> dumpDAGAfterPassesOpt(
+    "dump-graph-DAG-after-passes",
+    llvm::cl::desc("Debug option to export the Graph in DOT format after the "
+                   "listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(functionPassCat));
+
+llvm::cl::opt<bool> printPassesOpt(
+    "print-passes",
+    llvm::cl::desc("Print all of the passes run by the pass manager."),
+    llvm::cl::Optional, llvm::cl::cat(functionPassCat));
+
+llvm::cl::opt<int> stopAfterPassNumOpt(
+    "stop-passes-after-num",
+    llvm::cl::desc("Number of passes to run before preventing running any "
+                   "passes. Used for debugging."),
+    llvm::cl::init(-1), llvm::cl::cat(functionPassCat));
+
+/// Helper to check if \p otherStr is in \p strList.
+static bool listContainsString(llvm::ArrayRef<std::string> strList,
+                               llvm::StringRef otherStr) {
+  for (llvm::StringRef str : strList) {
+    if (str == otherStr) {
+      return true;
+    }
+  }
+  return false;
+}
+
+} // namespace
+
+int FunctionPass::globalPassCounter = 0;
+
+bool FunctionPass::runPrePass(Function *F, const CompilationContext &cctx,
+                              int count) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Starting Pass #" << count << ": " << this->getName().str()
+              << " on Function: \"" << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGBeforeAllPassesOpt ||
+      listContainsString(dumpDAGBeforePassesOpt, this->getName())) {
+    F->dumpDAG(F->getName().str() + "_PrePass_" + this->getName().str() + "_n" +
+               std::to_string(count) + ".dot");
+  }
+  if (verifyBeforeAllPassesOpt ||
+      listContainsString(verifyBeforePassesOpt, this->getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+bool FunctionPass::runPostPass(Function *F, const CompilationContext &cctx,
+                               int count) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Finished Pass #" << count << ": " << this->getName().str()
+              << " on Function: \"" << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGAfterAllPassesOpt ||
+      listContainsString(dumpDAGAfterPassesOpt, this->getName())) {
+    F->dumpDAG(F->getName().str() + "_PostPass_" + this->getName().str() +
+               "_n" + std::to_string(count) + ".dot");
+  }
+  if (verifyAfterAllPassesOpt ||
+      listContainsString(verifyAfterPassesOpt, this->getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+bool FunctionPass::run(Function *F, const CompilationContext &cctx) {
+  const int count = ++FunctionPass::globalPassCounter;
+
+  // If we've exceeded the number of passes to run then early exit.
+  if (stopAfterPassNumOpt >= 0 && count > stopAfterPassNumOpt) {
+    return false;
+  }
+
+  bool changed = runPrePass(F, cctx, count);
+  changed |= passImpl(F, cctx);
```

Comment:
I understand why you make clients override `passImpl`. But people who know e.g. LLVM are more used to overriding `run`. Maybe you can play with names and make clients override `run` and rename current `run`  into e.g. `runPass`, which would match nicely `runPrePass` and `runPostPass`? 

Nit: Another related question is: who should invoke pre and post functions? The pass itself or the PassManager? It can be done both ways. But if it is controlled by the pass, the PassManager has less insight into these three phases: pre, run, post and thus cannot e.g. log before/after each of these phases, etc. But this is probably not so important.